components.md

```diff
@@ -0,0 +1,98 @@
+# Components
+
+This page provides details on the main component/entity types that make up a deployment of the ECP.
+
+## Director
+A Director manages a part of or the entire setup, orchestrating the actors and observers according to the needs of the experiment.
+It can, among other things, 
+* issue commands to actors
+* request data from individual actors
+* read data from an observer
+
+Potentially a GUI could be attached here too.
+
+## Action
+An Action represents a method or function of the device represented by a Driver.
+It can be called with zero or more arguments.
+
+## Actor
+The Actor is a component which (typically) instantiates one or more Driver class and/or does some kind of processing.
+Actions which the actor can take could be anything from measuring on a regular basis ("acting" as a sensor), commanding hardware devices through the use of the driver to do X or Y, processing some data and re-emitting results (e.g. a PID controller).
+The actor is part of the communication network, that means it is listening for messages and maybe sending messages by itself.
+
+Potentially a GUI could be attached here too.
+
+The Actor would send commands to a Driver, instructing it to `set`/`get`/`call` a given Parameter/Action by name, with a value if appropriate, and receive replies.
+
+```{note}
+We could achieve more complete separation and make the Actor purely about processing _inputs_, doing some _operation_ and generating _outputs_.
+This could be either stateless (temperature conversion) or stateful (PID control).
+Maybe a better name would be "Processor" then.
+This means that _all_ device-related stuff stays compartmentalized in the Driver, and an Actor would address that Driver over the network (to fetch the Parameter values it is interested in).
```

Comment:
If we go through the network again, how do we call the recipient of the messages?

I understood that the actor is some software with two parts: a communication part and a "working" part, for example a driver. (although I'm unsure whether actor means the whole conglomerate or just the communication part)

Reply:
Not sure I understood correctly, let me try to illustrate
```mermaid
sequenceDiagram
    par
        Actor1->>Driver42: GET temp_K
        Driver42->>instr-on-COM1: "TEMP?\n"
        instr-on-COM1->>Driver42: "275.14\n"
        Driver42->>Actor1: 275.14
    and
        Actor1->>Driver11: GET temperature
        Note over Driver11: Did fetch a fresh value 23ms ago, using cached value
        Driver11->>Actor1: 300.1
    end
    Note over Actor1: computes average temperature, converts to degC
    Note over Actor1: stores/caches value maybe
    Actor1->>Observer1: "Average oven temperature" 14.47 degC
```
Is this clear?

Reply:
Yes

Reply:
So, agreed on putting the Device interaction all into Driver as I propose?

Reply:
> communicate in the same way as Actors and Observers - I do not know how to better separate those in mermaid though. Possibly it will be enough to "put a caption" to it in the sense of describing the abstractness of the arrows.

I noticed that, too. I can try to improve that (dashed arrows are possible I think).

> Students who never do anything if they did not get an input first (i.e. "process them") generally do not stay long.

:laughing: 

Yeah, I noticed that the boundaries felt weird/unclear, that's why asked for feedback on that. I don't think one or the other option is more "correct".

The motivation to separate this more, and to have the Driver participate in the network, was to

* get better separation of "business logic" (actor) and "HW comms layer" (driver)
* Have the Driver be a common interface for _all_ different instrument libraries to our protocol. I felt we should not put that into the Actor, too. 

You say
> An Actor is then a Component which participates in the ECP communication (contrary to the Driver), which may _use_ a Driver instance for its purposes
 
That "use" is the key -- how do we define that? How can an Actor _use_ a driver? What are a driver's capabilities? We are trying to create a not-Python-specific thing here, so we should not define this in terms of classes, properties, etc.
If we make the Driver a component, and define its interaction with the rest of ECP via _messages_, we can do that, people can code up a conforming implementation in their favourite language/style, and we can interoperate.

This was a bit stream-of-consciousness, but does that make the motivation clearer? How can I improve the text toward that (if you agree)?

Reply:
Let us continue in #17, to keep that discussion in one place, which can be found again.

Reply:
Resolved in #17 and with the updates, I think